In [2]:
import pandas as pd
import numpy as np
df= pd.read_csv('adult.csv')
#Sample rows
df.head(4)
# Dataset shape
print(f"Rows: {df.shape[0]}")
print(f"Columns: {df.shape[1]}")

#Dataset info
print(df.info())

#Looking for null values
print(df.isnull().sum())
print("Dataset Statistics")
print(df.describe(include="all"))



Rows: 32561
Columns: 15
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB
None
age               0
workclass         0
fnlwgt            0
educ

In [3]:
df= df.replace('?',np.nan)
df= df.replace(' ?',np.nan)
df.head()

age workclass  fnlwgt     education  education.num marital.status  \
0   90       NaN   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66       NaN  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship   race     sex  capital.gain  \
0                NaN  Not-in-family  White  Female             0   
1    Exec-managerial  Not-in-family  White  Female             0   
2                NaN      Unmarried  Black  Female             0   
3  Machine-op-inspct      Unmarried  White  Female             0   
4     Prof-specialty      Own-child  White  Female             0   

   capital.loss  hours.per.week native.country income  
0          4356              40  United-States  <=50K  
1          4356              18  United-States  <=50K  
2          4356              40  United-States  <=50K  
3          3900              40  United-States  <=50K  
4          3900              40  United-States  <=50K

In [4]:
print(df.isnull().sum())

age                  0
workclass         1836
fnlwgt               0
education            0
education.num        0
marital.status       0
occupation        1843
relationship         0
race                 0
sex                  0
capital.gain         0
capital.loss         0
hours.per.week       0
native.country     583
income               0
dtype: int64


In [5]:
cols_with_nan= df.columns[df.isnull().any()].tolist()
print(cols_with_nan)

['workclass', 'occupation', 'native.country']


In [6]:
#Preprocessing
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import seaborn as sns
import matplotlib.pyplot as plt


In [7]:
data= df.copy()
le= LabelEncoder()
data['race']= le.fit_transform(data['race'])
data['sex']= le.fit_transform(data['sex'])
print(data.head())

   age workclass  fnlwgt     education  education.num marital.status  \
0   90       NaN   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66       NaN  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship  race  sex  capital.gain  capital.loss  \
0                NaN  Not-in-family     4    0             0          4356   
1    Exec-managerial  Not-in-family     4    0             0          4356   
2                NaN      Unmarried     2    0             0          4356   
3  Machine-op-inspct      Unmarried     4    0             0          3900   
4     Prof-specialty      Own-child     4    0             0          3900   

   hours.per.week native.country income  
0              40  United-States  <=50K  
1             

In [8]:
plt.figure(figsize=(12,6))
sns.heatmap(data.isnull(),annot=True, cmap='Blues', cbar=False)
plt.title("Heatmap of Missing Values")
plt.show()

'plt.figure(figsize=(12,6))\nsns.heatmap(data.isnull(),annot=True, cmap=\'Blues\', cbar=False)\nplt.title("Heatmap of Missing Values")\nplt.show()'

In [10]:
data["workclass"] = data["workclass"].fillna("Unknown")
data["occupation"] = data["occupation"].fillna("Unknown")



In [11]:
data.head()

age workclass  fnlwgt     education  education.num marital.status  \
0   90   Unknown   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66   Unknown  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship  race  sex  capital.gain  capital.loss  \
0            Unknown  Not-in-family     4    0             0          4356   
1    Exec-managerial  Not-in-family     4    0             0          4356   
2            Unknown      Unmarried     2    0             0          4356   
3  Machine-op-inspct      Unmarried     4    0             0          3900   
4     Prof-specialty      Own-child     4    0             0          3900   

   hours.per.week native.country income  
0              40  United-States  <=50K  
1              18  United-States  <=50K  
2              40  United-States  <=50K  
3              40  United-States  <=50K  
4              40  United-States  <=50K

In [12]:
data["native.country"] = data["native.country"].fillna("Unknown")
data


age workclass  fnlwgt     education  education.num      marital.status  \
0       90   Unknown   77053       HS-grad              9             Widowed   
1       82   Private  132870       HS-grad              9             Widowed   
2       66   Unknown  186061  Some-college             10             Widowed   
3       54   Private  140359       7th-8th              4            Divorced   
4       41   Private  264663  Some-college             10           Separated   
...    ...       ...     ...           ...            ...                 ...   
32556   22   Private  310152  Some-college             10       Never-married   
32557   27   Private  257302    Assoc-acdm             12  Married-civ-spouse   
32558   40   Private  154374       HS-grad              9  Married-civ-spouse   
32559   58   Private  151910       HS-grad              9             Widowed   
32560   22   Private  201490       HS-grad              9       Never-married   

              occupation   relationship  race  sex  capital.gain  \
0                Unknown  Not-in-family     4    0             0   
1        Exec-managerial  Not-in-family     4    0             0   
2                Unknown      Unmarried     2    0             0   
3      Machine-op-inspct      Unmarried     4    0             0   
4         Prof-specialty      Own-child     4    0             0   
...                  ...            ...   ...  ...           ...   
32556    Protective-serv  Not-in-family     4    1             0   
32557       Tech-support           Wife     4    0             0   
32558  Machine-op-inspct        Husband     4    1             0   
32559       Adm-clerical      Unmarried     4    0             0   
32560       Adm-clerical      Own-child     4    1             0   

       capital.loss  hours.per.week native.country income  
0              4356              40  United-States  <=50K  
1              4356              18  United-States  <=50K  
2              4356              40  United-States  <=50K  
3              3900              40  United-States  <=50K  
4              3900              40  United-States  <=50K  
...             ...             ...            ...    ...  
32556             0              40  United-States  <=50K  
32557             0              38  United-States  <=50K  
32558             0              40  United-States   >50K  
32559             0              40  United-States  <=50K  
32560             0              20  United-States  <=50K  

[32561 rows x 15 columns]

In [14]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
# Group rare categories into 'Other'
rare_countries = data["native.country"].value_counts()[data["native.country"].value_counts() < 200].index
data["native.country"] = data["native.country"].replace(rare_countries, "Other")
data.head()

age workclass  fnlwgt     education  education.num marital.status  \
0   90   Unknown   77053       HS-grad              9        Widowed   
1   82   Private  132870       HS-grad              9        Widowed   
2   66   Unknown  186061  Some-college             10        Widowed   
3   54   Private  140359       7th-8th              4       Divorced   
4   41   Private  264663  Some-college             10      Separated   

          occupation   relationship  race  sex  capital.gain  capital.loss  \
0            Unknown  Not-in-family     4    0             0          4356   
1    Exec-managerial  Not-in-family     4    0             0          4356   
2            Unknown      Unmarried     2    0             0          4356   
3  Machine-op-inspct      Unmarried     4    0             0          3900   
4     Prof-specialty      Own-child     4    0             0          3900   

   hours.per.week native.country income  
0              40  United-States  <=50K  
1              18  United-States  <=50K  
2              40  United-States  <=50K  
3              40  United-States  <=50K  
4              40  United-States  <=50K

In [15]:
#One-Hot Encoding
data = pd.get_dummies(data, columns=["marital.status"], drop_first=True)
data

age workclass  fnlwgt     education  education.num         occupation  \
0       90   Unknown   77053       HS-grad              9            Unknown   
1       82   Private  132870       HS-grad              9    Exec-managerial   
2       66   Unknown  186061  Some-college             10            Unknown   
3       54   Private  140359       7th-8th              4  Machine-op-inspct   
4       41   Private  264663  Some-college             10     Prof-specialty   
...    ...       ...     ...           ...            ...                ...   
32556   22   Private  310152  Some-college             10    Protective-serv   
32557   27   Private  257302    Assoc-acdm             12       Tech-support   
32558   40   Private  154374       HS-grad              9  Machine-op-inspct   
32559   58   Private  151910       HS-grad              9       Adm-clerical   
32560   22   Private  201490       HS-grad              9       Adm-clerical   

        relationship  race  sex  capital.gain  capital.loss  hours.per.week  \
0      Not-in-family     4    0             0          4356              40   
1      Not-in-family     4    0             0          4356              18   
2          Unmarried     2    0             0          4356              40   
3          Unmarried     4    0             0          3900              40   
4          Own-child     4    0             0          3900              40   
...              ...   ...  ...           ...           ...             ...   
32556  Not-in-family     4    1             0             0              40   
32557           Wife     4    0             0             0              38   
32558        Husband     4    1             0             0              40   
32559      Unmarried     4    0             0             0              40   
32560      Own-child     4    1             0             0              20   

      native.country income  marital.status_Married-AF-spouse  \
0      United-States  <=50K                             False   
1      United-States  <=50K                             False   
2      United-States  <=50K                             False   
3      United-States  <=50K                             False   
4      United-States  <=50K                             False   
...              ...    ...                               ...   
32556  United-States  <=50K                             False   
32557  United-States  <=50K                             False   
32558  United-States   >50K                             False   
32559  United-States  <=50K                             False   
32560  United-States  <=50K                             False   

       marital.status_Married-civ-spouse  \
0                                  False   
1                                  False   
2                                  False   
3                                  False   
4                                  False   
...                                  ...   
32556                              False   
32557                               True   
32558                               True   
32559                              False   
32560                              False   

       marital.status_Married-spouse-absent  marital.status_Never-married  \
0                                     False                         False   
1                                     False                         False   
2                                     False                         False   
3                                     False                         False   
4                                     False                         False   
...                                     ...                           ...   
32556                                 False                          True   
32557                                 False                         False   
32558                                 False                         False   
32559            

In [16]:
X = data.drop(columns=["income", "fnlwgt"])
y= data['income']
model= LogisticRegression()

In [17]:
print(X.dtypes)

age                                      int64
workclass                               object
education                               object
education.num                            int64
occupation                              object
relationship                            object
race                                     int64
sex                                      int64
capital.gain                             int64
capital.loss                             int64
hours.per.week                           int64
native.country                          object
marital.status_Married-AF-spouse          bool
marital.status_Married-civ-spouse         bool
marital.status_Married-spouse-absent      bool
marital.status_Never-married              bool
marital.status_Separated                  bool
marital.status_Widowed                    bool
dtype: object


In [18]:
num_cols = X.select_dtypes(include=["int64", "float64"]).columns
cat_cols = X.select_dtypes(include=["object"]).columns
print(num_cols)
print(cat_cols)



Index(['age', 'education.num', 'race', 'sex', 'capital.gain', 'capital.loss',
       'hours.per.week'],
      dtype='object')
Index(['workclass', 'education', 'occupation', 'relationship',
       'native.country'],
      dtype='object')


In [19]:
#Scale numeric columns
from sklearn.preprocessing import StandardScaler

num_cols = ['age', 'education.num', 'capital.gain', 'capital.loss', 'hours.per.week']
scaler = StandardScaler()
X[num_cols] = scaler.fit_transform(X[num_cols])
print(data[num_cols])

       age  education.num  capital.gain  capital.loss  hours.per.week
0       90              9             0          4356              40
1       82              9             0          4356              18
2       66             10             0          4356              40
3       54              4             0          3900              40
4       41             10             0          3900              40
...    ...            ...           ...           ...             ...
32556   22             10             0             0              40
32557   27             12             0             0              38
32558   40              9             0             0              40
32559   58              9             0             0              40
32560   22              9             0             0              20

[32561 rows x 5 columns]


In [22]:
X = pd.get_dummies(X, drop_first=True)



In [23]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)


In [24]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)



LogisticRegression(max_iter=1000)

This output shows the predicted income category for each test sample. The model predicts whether a person earns more than $50K or  less   than or equal to $50K per year based on features like age, education, workclass, and hours per week. Each value in the array corresponds to one record in the test dataset.

In [26]:
y_pred = model.predict(X_test)
y_pred



array(['<=50K', '<=50K', '<=50K', ..., '>50K', '<=50K', '<=50K'],
      dtype=object)

In [27]:
from sklearn.metrics import accuracy_score

print("Accuracy:", accuracy_score(y_test, y_pred))



Accuracy: 0.8452326116996776


In [28]:
from sklearn.metrics import confusion_matrix

print(confusion_matrix(y_test, y_pred))



[[4621  355]
 [ 653  884]]


4621 people correctly predicted as <=50K

884 people correctly predicted as >50K

355 people wrongly predicted as >50K

653 people wrongly predicted as <=50K